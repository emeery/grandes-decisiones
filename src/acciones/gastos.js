import uuid from 'uuid';

export const crearGasto = ({
    descripcion = '', 
    nota = '',
    monto = 0,
    date = 0
 }) => ({
    type: 'SUMA_GASTO',
    gasto: {
        id: uuid(),
        descripcion,
        nota,
        monto,
        date
    }
});
// 
export const removerGasto = ({ id } = {}) => ({
    type: 'REMUEVE_GASTO',
    id
});
// editar descripcion o monto, 
export const editarGasto = (id, update) => ({
    type: 'EDITA_GASTO',
    id,
    update
});