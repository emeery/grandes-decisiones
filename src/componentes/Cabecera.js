import React from 'react';
import {NavLink} from 'react-router-dom';

// <NavLink to="/ayuda" activeClassName='is-active'>Ayuda</NavLink>
const Cabecera = () => (
    <header>
    <h1>Libros de Notas.</h1>
    <h3>La pieza faltante.</h3>
    <NavLink to="/" activeClassName='is-active' exact={true} >Escritorio</NavLink>
    <NavLink to="/crealo" activeClassName='is-active'>Crear Gastos</NavLink>
    
    </header>

); export default Cabecera;

