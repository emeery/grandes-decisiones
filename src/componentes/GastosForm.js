import React from 'react';
import moment from 'moment';
import { SingleDatePicker} from 'react-dates'; 

// const ahora = moment();
// console.log(ahora.format('MMM Do, YYYY') );

class GastosForm extends React.Component {
    constructor(props) {
        super( props );
        this.state = {
        descripcion: props.gasto ?  props.gasto.descripcion : '' , 
        nota: props.gasto ? props.gasto.nota : '' , 
        monto: props.gasto ? (props.gasto.monto / 100 ).toString() : '', 
        date: props.gasto ? moment(props.gasto.date) : moment(),
        calendarFocused : false,
        error: ''
        };
    }
    descripcionCambio = (e) => {
        const descripcion = e.target.value; 
        this.setState(() => ({ descripcion })); 
    };
    // persistncia ();
    notaCambio = (e) => {
        const nota = e.target.value;
        this.setState(()=> ({ nota }));
    };
    montoCambio = (e) => {
        const monto = e.target.value;
        if (!monto || monto.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ monto }));
        }
    };
    onDatesChange = (date) => {
        if(date) {
            this.setState(()=> ({ date }));
        }
    };
    // focused - requerido
    onFocusChange = ({focused}) => {
        this.setState(() => ({ calendarFocused: focused }));
    };
    onSubmit = (e) => {
        e.preventDefault();
        if (!this.state.descripcion || !this.state.monto) {
            this.setState(() => ({error: 'ingresa los 2 valores'}));
        } else {
            this.setState(() => ({error: ''}));
            this.props.onSubmit({
                descripcion: this.state.descripcion,
                monto: parseFloat(this.state.monto, 10) * 100,
                date: this.state.date.valueOf(), // nmero regular que representa esa fecha
                nota: this.state.nota
            });
        }
    };
    render() {
        return( <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmit}>      
            <input 
            type='text'
            placeholder='Descripcion'
            autoFocus
            value={this.state.descripcion}
            onChange={this.descripcionCambio} 
            />
            <input
            type='text'
            placeholder='Monto'
            value={this.state.monto}
            onChange={this.montoCambio}
            />
            
            <SingleDatePicker
            date={this.state.date}
            onDateChange={this.onDatesChange}
            focused={this.state.calendarFocused} 
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={()=>false}
            />
            
            <textarea
            placeholder='agrega una nota para tu gasto'
            value={this.state.nota}
            onChange={this.notaCambio}
            >
            </textarea>

            <button>Agrega Gasto</button>
        </form>    
        </div>);
    }
}; export default GastosForm;


// crear playlist.
// nombrePLaylist /* -string, nombre 1/100*/
// <button /*  elegir imagen */
// descripcion /* -string estos tres 0/300 */
// <button /* cancelar */
// <button /* crear */
