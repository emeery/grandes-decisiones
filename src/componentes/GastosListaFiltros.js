import React from 'react';
import {connect} from 'react-redux';
import {DateRangePicker} from 'react-dates'; 

import {
    establecerTextoFiltros, 
    ordenarPorMonto, 
    ordenarPorFecha,
    establecerFechaInicio, 
    establecerFechaFin
} from '../acciones/filtros';

class GastosListaFiltros extends React.Component {
    state = {
        calendarFocused: null
    };
    onDatesChange = ({startDate, endDate}) => {
        this.props.dispatch(establecerFechaInicio(startDate) );
        this.props.dispatch(establecerFechaFin(endDate) );
    };
    onFocusChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused }));
    }
    render() {
        return  ( 
        <div>
        
        <input 
        type='text' 
        value={this.props.filtros.texto}
        onChange={(e) => {
            this.props.dispatch(establecerTextoFiltros(e.target.value));
        }} />
        
        <select 
        value={this.props.filtros.ordenarPor}
        onChange={(e) => {
            if (e.target.value === 'fecha') {
                this.props.dispatch(ordenarPorFecha() );
            } else if (e.target.value === 'monto') {
                this.props.dispatch(ordenarPorMonto() );
            }
        }}
        >
        <option value='fecha'>
            fecha </option>
        <option value='monto'>
            monto </option>
        
        </select>
        <DateRangePicker
        startDate={this.props.filtros.startDate}
        endDate={this.props.filtros.endDate}
        onDatesChange={this.onDatesChange}
        focusedInput={this.state.calendarFocused}
        onFocusChange={this.onFocusChange}
        showClearDates={true}
        numberOfMonths={1}
        isOutsideRange={() => false}
        />
    </div> );
    }
};

const mapStateToProps = (state) => {
    return { filtros: state.filtros };
};

export default connect(mapStateToProps)(GastosListaFiltros);