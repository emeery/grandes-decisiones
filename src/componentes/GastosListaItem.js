import React from 'react';
import { Link } from 'react-router-dom';
//import {connect } from 'react-redux';
// import {removerGasto} from '../acciones/gastos'; connect 

// destructuracion dispatch
        
//dispatch, id, descripcion, monto, fecha
const GastosListaItem = ({
    id, descripcion, monto, date}) => (
    <div>
        <Link to={`/edita/${id}`}>
        <h3> {descripcion} </h3>
        </Link>
        <h2> {monto} - {date}</h2>
       
    </div>
);
export default GastosListaItem;

// const GastosListaItem = ({ 
//      dispatch, id, descripcion, monto, fecha}) => (
//      <div>
//         <h3> descripcion </h3>
//         <h2>fecha: {fecha} - monto: {monto}</h2>
//         <button onClick={() => {
//             dispatch(removeGasto({ id }));
//         }}
//         > remover </button>
//     </div>
// );
// export default connect()(GastosListaItem);

// <h2> monto fecha </h2>
// <button onClick={() => {
//     dispatch(removerGasto({ id }));
// }}
// > remover </button>