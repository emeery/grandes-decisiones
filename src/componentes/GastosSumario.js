import React from 'react';
import {connect} from 'react-redux';
import numeral from 'numeral';
import seleccionaGastos from '../selectores/gastos'; // creadoEnElMomento
import seleccionaGastosTotal from '../selectores/gastos-totales';

export const GastosSumario = ({ gastosContador, gastosTotal}) => {
    //const sinGastos = gastosContador === 0 ? 'no hay gastos' : gastosContador;
    const gastosPalabra = gastosContador === 1 ? 'Gasto' : 'Gastos';
    //gastosPalabra === 0 ? '' : gastosContador;
    const formatoGastosTotal = numeral(gastosTotal / 100).format('$0,0.00'); 
    
    return(<div>
        <h2> 
        Viendo -
        {gastosContador}
        {gastosPalabra} -
        Total - 
        {formatoGastosTotal}</h2>
        </div>);
};

const mapStatetoProps = (state) => {
    const gastosVisibles = seleccionaGastos(state.gastos, state.filtros );
    return {
        gastosContador: gastosVisibles.length,
        gastosTotal: seleccionaGastosTotal(gastosVisibles)
    };
};

export default connect(mapStatetoProps)(GastosSumario)