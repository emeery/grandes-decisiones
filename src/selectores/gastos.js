import moment from 'moment';
// valida las fechas y el texto /
// seleccionaGasto
export default (gastos, {texto, ordenarPor, startDate, endDate}) => {
    return gastos.filter((gasto) => {
        const creadoEnElMomento = moment(gasto.date);
        const fechaInicioMatch = startDate ? startDate.isSameOrBefore(creadoEnElMomento, 'day') : true;
        const fechaFinMatch = endDate ? endDate.isSameOrAfter(creadoEnElMomento, 'day'): true;
        const textoMatch = gasto.descripcion.toLowerCase().includes(texto);
        return textoMatch && fechaInicioMatch && fechaFinMatch;
    }).sort(( a , b )=> {
        if (ordenarPor === 'date') {
            return a.date < b.date ? 1 : -1;
        } else if (ordenarPor === 'monto') {
            return a.monto < b.monto ? 1 : -1;
        }
    });
};

// The selectors are functions that take the status of Redux 
// as an argument and return some data to pass to the component.

// const fechaInicioMatch = typeof fechaInicio !== 'number' || gasto.fecha >= fechaInicio;
// const fechaFinMatch = typeof fechaFin !== 'number' || gasto.fecha <= fechaFinMatch;
