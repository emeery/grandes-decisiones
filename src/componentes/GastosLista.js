import React from 'react';
import { connect } from 'react-redux';
import GastosListaItem from './GastosListaItem';
import seleccionaGastos from '../selectores/gastos';
import gastos from '../test/accesorios/gastos';

// conecta tu componente a la tienda redux
export const GastosLista = (props) => (
    <div>
        {
            props.gastos.length === 0 ? (
            <p> No hay Gastos ! </p>
        ) : (
            props.gastos.map((gastitos)=>{
                return <GastosListaItem 
                key={gastitos.id} 
                {...gastitos}
                />;
            })
        )}
        
    </div>
); 
// definimos lo que queremos de la tienda
// y definimos el componnte GastosLista 

const mapStatetoProps = (state) => {
    return { 
        gastos: seleccionaGastos(state.gastos, state.filtros)
    };
};
export default connect(mapStatetoProps)(GastosLista);
