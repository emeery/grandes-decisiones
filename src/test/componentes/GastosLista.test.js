import React from 'react';
import { shallow } from 'enzyme';
import {GastosLista} from '../../componentes/GastosLista';
import gastos from '../accesorios/gastos';

test('la lista de gastos con algunos gastos', ()=>{
    const envoltura = shallow(
        <GastosLista gastos={gastos} />
    );
    expect(envoltura).toMatchSnapshot();
});

test('los gastosl lista sin contenido',()=>{
    const envoltura = shallow(
        <GastosLista gastos={[]} />
    );
    expect(envoltura).toMatchSnapshot();
});

