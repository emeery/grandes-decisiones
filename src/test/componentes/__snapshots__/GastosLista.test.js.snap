// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`la lista de gastos con algunos gastos 1`] = `
<div>
  <GastosListaItem
    date={0}
    descripcion="goma"
    id="1"
    monto={195}
    nota=""
  />
  <GastosListaItem
    date={-345600000}
    descripcion="renta"
    id="2"
    monto={109500}
    nota=""
  />
  <GastosListaItem
    date={345600000}
    descripcion="tarjeta credito"
    id="3"
    monto={4500}
    nota=""
  />
</div>
`;

exports[`los gastosl lista sin contenido 1`] = `
<div>
  <p>
     No hay Gastos ! 
  </p>
</div>
`;
